packages/tpetra/core/src/Tpetra_RowMatrixTransposer_def.hpp

```diff
@@ -120,62 +120,131 @@ createTransposeLocal (const Teuchos::RCP<Teuchos::ParameterList> &params)
   RCP<const crs_matrix_type> crsMatrix =
     rcp_dynamic_cast<const crs_matrix_type> (origMatrix_);
-  if (crsMatrix == Teuchos::null) {
+
+  RCP<crs_matrix_type> transMatrixWithSharedRows;
+  if (crsMatrix != Teuchos::null) {
+    // The matrix is a CrsMatrix
+    using local_matrix_type = typename crs_matrix_type::local_matrix_type;
+    using row_map_type = typename local_matrix_type::row_map_type::non_const_type;
+    using index_type = typename local_matrix_type::index_type::non_const_type;
+    using values_type = typename local_matrix_type::values_type::non_const_type;
+    using execution_space = typename local_matrix_type::execution_space;
+
+    // Determine how many nonzeros there are per row in the transpose.
+    auto lclMatrix = crsMatrix->getLocalMatrix();
+    auto lclGraph  = lclMatrix.graph;
+
+    using range_type = Kokkos::RangePolicy<LO,execution_space>;
+
+    // Determine how many nonzeros there are per row in the transpose.
+    row_map_type t_rows("transpose_rows", numLocalCols+1);
+    Kokkos::parallel_for("compute_number_of_indices_per_column", range_type(0, numLocalRows),
+      KOKKOS_LAMBDA(const LO row) {
+        auto rowView = lclGraph.rowConst(row);
+        auto length  = rowView.length;
+
+        for (decltype(length) colID = 0; colID < length; colID++) {
+          auto col = rowView(colID);
+          Kokkos::atomic_fetch_add(&t_rows[col], 1);
+        }
+    });
+
+    // Compute offsets
+    Kokkos::parallel_scan("compute_transpose_row_offsets", range_type(0, numLocalCols+1),
+      KOKKOS_LAMBDA(const LO i, LO& update, const bool& final_pass) {
+        const LO val = t_rows(i);
+        if (final_pass)
+          t_rows(i) = update;
+        update += val;
+      });
+
+    row_map_type offsets("transpose_row_offsets_aux", numLocalCols+1);
+    Kokkos::deep_copy(offsets, t_rows);
+
+    index_type  t_cols("transpose_cols", numLocalNnz);
+    values_type t_vals("transpose_vals", numLocalNnz);
+    Kokkos::parallel_for("compute_transposed_rows", range_type(0, numLocalRows),
+      KOKKOS_LAMBDA(const LO row) {
+        auto rowView = lclMatrix.rowConst(row);
+        auto length  = rowView.length;
+
+        for (decltype(length) colID = 0; colID < length; colID++) {
+          auto col = rowView.colidx(colID);
+
+          LO insert_pos = Kokkos::atomic_fetch_add(&offsets[col], 1);
+
+          t_cols[insert_pos] = row;
+          t_vals[insert_pos] = rowView.value(colID);
+        }
+    });
+
+    local_matrix_type lclTransposeMatrix("transpose", numLocalCols, numLocalRows, numLocalNnz, t_vals, t_rows, t_cols);
+
+    transMatrixWithSharedRows =
+        rcp (new crs_matrix_type (lclTransposeMatrix,
+                                  origMatrix_->getColMap (), origMatrix_->getRowMap (),
+                                  origMatrix_->getRangeMap (), origMatrix_->getDomainMap ()));
+  } else {
+    // FIXME: are we ever here? There is no RowMatrix constructor, we seem to
```

Comment:
Huh, this is a good point.  It's not really a "RowMatrixTransposer" then ;-) .  It's up to you whether you would like to purge this code path.

Reply:
It seems that the proper solution would instead be to introduce a `RowMatrix` constructor, and store the row matrix that way. This way, it expands the applicability of the transposer. However, it's not clear to me if that is tested anywhere.